s2/buffer_pool.go

```diff
@@ -0,0 +1,64 @@
+//  Copyright (c) 2023 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s2
+
+// GeoBufferPool represents a pool of buffers ranging from a given
+// max size to a min size in steps of 2. It uses a lazy approach only allocating
+// the buffers when it is needed.
+
+type GeoBufferPool struct {
+	buffers [][]byte
+	maxSize int
+	minSize int
+}
+
+func NewGeoBufferPool(maxSize int, minSize int) *GeoBufferPool {
+	// Calculating the number of buffers required. Assuming that
+	// the value of minSize is correct, the buffers will be of size
+	// minSize, 2 * minSize, 4 * minSize and so on till it is less
+	// than or equal to the maxSize. If it is not equal to maxSize,
+	// then a suitable value less than maxSize will be set as maxSize
+	length := 0
+	temp := minSize
+	for temp <= maxSize {
+		length = length + 1
+		temp = temp * 2
+	}
+	maxSize = temp / 2
+
+	return &GeoBufferPool{
+		buffers: make([][]byte, length),
+		maxSize: maxSize,
+		minSize: minSize,
+	}
+}
+
+func (b *GeoBufferPool) Get(size int) ([]byte) {
+	bufSize := b.maxSize
+
+	for i := range b.buffers {
+		if bufSize <= size {
```

Comment:
this is my first parse of the PR, but if the caller asks for a buffer size let's say 1023 (Get(1023)), then would you end up returning a buffer of size 512 instead of 1024?  like isn't the condition for the buffer size that's closes to the size but greater than size? (so that we have allocated enough space and there won't be out of bounds access etc). please let me know if this makes sense

Reply:
The condition is the largest buffer that is less than or equal to the buffer size. So 2047 should return 1024.

Reply:
@Likith101 I kind of agree with what @Thejas-bhat  said.
For a Get(100) I am getting a []byte of size 96. 
I will have to use 2 read system calls to read 100 bytes content from a file ( If I am using a buffer of size 96).
In the buffer pool next bigger item is 96*2=196 bytes item and we should return that to user.

So essentially algorithm should be, 
"smallest buffer greater than equal to requested size"

Reply:
But in case we don't have a buffer which can satisfy the above requirement.
Then in that case, we will return the largest available buffer

Reply:
```
// returns smallest available buffer of size >= requestd size.
// If no buffer is available, returns a buffer of max size.
func (b *GeoBufferPool) Get(size int) []byte {
	currBuffSize := b.minSize
	for i := 0; i < len(b.buffers); i++ {
		if currBuffSize >= size {
			if b.buffers[i] == nil {
				b.buffers[i] = make([]byte, currBuffSize)
			}
			return b.buffers[i]
		}

		currBuffSize = currBuffSize * 2
	}

	// return available buffer of max size
	if b.buffers[len(b.buffers)-1] == nil {
		b.buffers[len(b.buffers)-1] = make([]byte, b.maxSize)
	}

	return b.buffers[len(b.buffers)-1]
}
```

Reply:
Great points by both @Thejas-bhat and @moshaad7 .

I think it makes sense to use a buffer of size **equal to or greater than** what needs to be read up **_until_** the cap (max buffer size in pool) to avoid multiple reads whenever possible.